common/djangoapps/third_party_auth/tests/test_decorators.py

```diff
@@ -57,3 +70,67 @@ def test_feature_flag_off(self, url):
             request = self.construct_request(url)
             response = mock_view(request)
             self.assertEqual(response['X-Frame-Options'], 'DENY')
+
+
+@unittest.skipIf(
+    'third_party_auth' not in settings.INSTALLED_APPS,
+    'third_party_auth is not currently installed in CMS'
+)
+@ddt.ddt
+class TestTpaHintSessionEndingDecorator(TestCase):
+
+    url = '/protected_view'
+
+    def setUp(self):
+        super(TestTpaHintSessionEndingDecorator, self).setUp()
+        self.enable_saml()
+        self.factory = RequestFactory()
+
+    def get_user_mock(self, authenticated=False):
+        user = MagicMock()
+        user.is_authenticated.return_value = authenticated
+        return user
+
+    def get_request_mock(self, authenticated=False, hinted=False, done=False):
+        user = self.get_user_mock(authenticated)
+        params = {}
+        if hinted:
+            params['tpa_hint'] = 'saml-realprovider'
+        if done:
+            params['session_cleared'] = 'yes'
+        url = '{}?{}'.format(self.url, urlencode(params))
+        request = self.factory.get(url)
+        request.user = user
+        return request
+
+    def create_provider(self, protected=False):
+        self.configure_saml_provider(
+            enabled=True,
+            name="Fancy real provider",
+            idp_slug="realprovider",
+            backend_name="tpa-saml",
+            drop_existing_session=protected,
+        )
+
+    @ddt.unpack
+    @ddt.data(
+        (True, True, False, False, False),
+        (True, True, True, False, True),
+        (True, True, True, True, False),
+        (True, True, False, True, False),
+        (False, True, True, False, False),
+        (True, False, True, False, False),
```

Comment:
Nit: This grid of booleans makes this test case a little hard to follow. I would suggest splitting this up into two or more test functions, and instead of making 'done' a parameter and mocking the requests, use `self.client()` and real requests to ensure that the redirects work (by actually following them) and don't loop infinitely. That would also let you avoid encoding `session_cleared` into the test case, which is more of an implementation detail than a part of the interface.

That said, consider this feedback optional. You have something that works, and it's fine.